runtime/interpreter/value.go

```diff
@@ -13750,6 +13750,12 @@ func (v *SomeValue) NeedsStoreTo(address atree.Address) bool {
 }
 
 func (v *SomeValue) IsResourceKinded(interpreter *Interpreter) bool {
+	// Assumption: inner-value of an optional could only be nil
+	// if it is a resource and has been invalidated/destroyed.
+	if v.value == nil {
+		return true
+	}
```

Comment:
Had to add this, since now inner-value could be nil, and could crash while trying to check if the value is a resource.
cc: @turbolent

Reply:
I see, thank you for pointing this out 👍 

Instead of assuming the value must be nil here because of the resource invalidation, try replacing it with the checking code that is also performed in the other functions:

```go
	if interpreter.invalidatedResourceValidationEnabled {
		// TODO: provide proper location range
		v.checkInvalidatedResourceUse(ReturnEmptyLocationRange)
	}
```

Reply:
hmm, that way it losses the line number info. We can create and pass a locationGetter with proper line number info here, but that would again add the overhead.

Maybe I'll special-case the optional-values at the resource checking methods, instead of here

Reply:
Refactored in d892eba22c938f5109d31ec456827ddcb896f94b, as I suggested above, so that we can keep both the performance as well as line number info where-ever possible.

Reply:
Nice, I will have a look 👍